impl/src/main/java/org/ehcache/internal/events/CacheEventNotificationListenerServiceProviderImpl.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ehcache.internal.events;
+
+import org.ehcache.events.CacheEventNotificationListenerServiceProvider;
+import org.ehcache.events.CacheEventNotificationService;
+import org.ehcache.events.CacheEventNotificationServiceImpl;
+import org.ehcache.events.DisabledCacheEventNotificationService;
+import org.ehcache.spi.ServiceProvider;
+import org.ehcache.spi.service.ServiceConfiguration;
+import org.ehcache.spi.service.ThreadPoolsService;
+
+/**
+ * @author palmanojkumar
+ *
+ */
+public class CacheEventNotificationListenerServiceProviderImpl implements CacheEventNotificationListenerServiceProvider {
+
+  private ServiceConfiguration<?> config;
+  private ServiceProvider serviceProvider;
```

Comment:
I think there's still some confusion around the services layer. My original understanding was that all the services are to be initialized eagerly during cache manager initialization and not used anymore once everything is up and running.

I now realize this is wrong and that the services could be called at any time after the cache manager is initialized, simply by creating new caches with CacheManager.createCache(). I therefore agree that marking such fields as volatile is the best thing to do, but there are other services with such problem that would need to be fixed too. I'll file a bug about them.

Regarding this PR, I agree with Alex that those fields should be volatile.